guides/manual-guide-to-rings.md

```diff
@@ -0,0 +1,323 @@
+# A manual guide to utilizing rings
+
+### Introduction
+
+This guide presumes that you have set up your `spk` project, and installed all
+necessary pipelines created via `spk project init` (the lifecycle pipeline),
+`spk service create` (build update hld pipeline), and `spk hld init` (manifest
+generation pipeline), and followed the
+[guidelines for creating helm charts](./building-helm-charts-for-spk).
+
+In `spk`, we offer the concept of a `ring` - a way to route inbound traffic to
+_revisions_ of a service on a Kubernetes cluster via request headers. For
+example, if an inbound request is decorated with the `Ring: dev` header, the
+request is routed to the `dev` revision of a service. Similarly, if a request is
+decorated with the `Ring: prod` header, the request is routed to the `prod`
+revision of a service.
+
+### Service Revisions and Git
+
+Service revisions are built and deployed by committing to bedrock.yaml tracked
+git branches in the form of a `ring`. A `ring` maps one to one with a service
+revision, which maps directly onto a git branch.
+
+Let's think about a git repository containing three branches, `dev`, `qa`, and
+`prod`. Each of these branches contain variations of an application's source
+code. `dev` containing the in-progress feature work, `qa` containing stable, but
+under-test feature work, and `prod`, containing live/production ready features.
+
+A `bedrock.yaml` for this service might look like:
+
+```yaml
+rings:
+  dev:
+    isDefault: true
+  qa:
+  prod:
+services:
+  ./:
+    displayName: "fabrikam"
+    helm:
+      chart:
+        branch: master
+        git: "https://dev.azure.com/fabrikam/frontend/_git/charts"
+        path: frontend
+    k8sBackend: "fabrikam-k8s-svc"
+    k8sBackendPort: 80
+    middlewares: []
+    pathPrefix: "fabrikam-service"
+    pathPrefixMajorVersion: "v1"
+variableGroups:
+  - fabrikam-vg
+```
+
+Each of the three entries under the `rings` object map one to one with a git
+branch in the application source repository. When this `bedrock.yaml` is
+deployed, and the container images for each of the `rings` are built and pushed
+for the `fabrikam` service, your cluster will have three running revisions of
+the `fabrikam` application - one each for `dev`, `qa`, and `prod`. A user is
+then able to invoke each revision by making HTTP requests to a single endpoint,
+decorated with the proper header for each Ring.
+
+### How deploying service revisions with rings will work
+
+Refer to the `bedrock.yaml` above, with the following rings, and thus git
+branches:
+
+```
+- dev
+- prod
+- qa
+```
+
+A user wants to add a ring `test-new-homepage`, they first ensure that the git
+branch exists:
+
+- `git checkout dev -b test-new-homepage`
+
+and then invoke the relevant `spk` command to add a `ring`:
```

Comment:
This workflow is a bit confusing. This is suggesting modifying `bedrock.yaml` in this new branch, which it should be done in the branch, usually `master`, that the hld-lifecycle pipeline triggers off of.

Reply:
cleared this up